src/index.ts

```diff
@@ -242,25 +389,55 @@ export function register (opts: RegisterOptions = {}): Register {
  * Create TypeScript compiler instance.
  */
 export function create (options: CreateOptions = {}): Register {
+  return createInternal(new OptionsHelper({
+    explicitOptions: options
+  }))
+}
+
+function createInternal (optionsHelper: OptionsHelper): Register {
+  let options = optionsHelper.merge()
   const ignoreDiagnostics = [
     6059, // "'rootDir' is expected to contain all source files."
     18002, // "The 'files' list in config file is empty."
     18003, // "No inputs were found in config file."
     ...(options.ignoreDiagnostics || [])
   ].map(Number)
 
-  const ignore = options.skipIgnore ? [] : (options.ignore || ['/node_modules/']).map(str => new RegExp(str))
-
   // Require the TypeScript compiler and configuration.
-  const cwd = options.dir ? resolve(options.dir) : process.cwd()
-  const isScoped = options.scope ? (fileName: string) => relative(cwd, fileName).charAt(0) !== '.' : () => true
+
+  /**
+   * Compute options that must be computed before *and* after loading tsconfig
+   * They are required to successfully parse tsconfig, but might be changed by
+   * ts-node options specified in the config file.
+   */
+  function recomputedOptions () {
+    const cwd = options.dir ? resolve(options.dir) : process.cwd()
+    const isScoped = options.scope ? (fileName: string) => relative(cwd, fileName).charAt(0) !== '.' : () => true
+    const compiler = require.resolve(options.compiler || 'typescript', { paths: [cwd, __dirname] })
+    const ts: typeof _ts = require(compiler)
+    const readFile = options.readFile || ts.sys.readFile
+    const fileExists = options.fileExists || ts.sys.fileExists
+    return { cwd, isScoped, compiler, ts, readFile, fileExists }
+  }
+
+  // compute enough options to read the config file
+  let { cwd, isScoped, compiler, ts, fileExists, readFile } = recomputedOptions()
```

Comment:
I don't think we should be doing `recomputedOptions` like this, it feels wasteful. There's only two options here (`isScoped` and `compiler`) that can change from `tsconfig.json`. `isScoped` could be populated _after_ `tsconfig.json` is loaded, therefore this is really only here to load `compiler` again.

Reply:
I'll move `cwd` and `isScoped` outside of this function.

The other 4 -- `compiler`, `ts`, `readFile`, and `fileExists` -- are all needed both before and after `readConfig()`.  Should I rename the function to something like `loadCompiler`?  I can also get rid of the destructuring if you want; just assign straight to local `let` variables.

Reply:
Yeah, for `readFile` and `fileExists`, we should probably just fallback on `ts.sys` within the `readConfig` function directly - I don't think those two functions are actually used consistently when I looked at the code, so it wouldn't be a loss to lose the functionality altogether.

Reply:
I see what you mean.  `ts.parseJsonConfigFileContent` accepts a `ParseConfigHost`, which has `readFile`, `fileExists`, and `readDirectory`.  We can add a `readDirectory` option to `RegisterOptions`, and then we can consistently use `readFile` and `fileExists` in a way that makes sense.

Who is using these options, and what do they use them for?

Reply:
It might be cleaner if ts-node accepted a `*Host`-style object instead of `readFile` and `fileExists`.  That way consumers of the API could override any of the various fs-related methods on `ts.sys`.  Looks like there are a bunch, like `directoryExists`, that are being used by ts-node but not overridable via the API.